pkg/controller/analytics/analytics_controller.go

```diff
@@ -0,0 +1,326 @@
+				return ctrl.Result{}, err
+			}
+			for _, u := range ul.Items {
+				m, ok, err := unstructured.NestedStringMap(u.Object, "spec", "selector", "matchLabels")
+				if !ok || err != nil {
+					return ctrl.Result{}, fmt.Errorf("%s not supported", gvr.String())
+				}
+				matchLabels := map[string]string{}
+				for k, v := range m {
+					matchLabels[k] = v
+				}
+				if match(rs.LabelSelector, matchLabels) {
+					us = append(us, u)
+					ownerNames = append(ownerNames, u.GetName())
+				}
+			}
+		}
+
+		for i := range us {
+			k := objRefKey(rs.Kind, rs.APIVersion, ownerNames[i])
+			if _, exists := identities[k]; !exists {
+				identities[k] = ObjectIdentity{
+					Namespace:  req.Namespace,
+					Name:       ownerNames[i],
+					Kind:       rs.Kind,
+					APIVersion: rs.APIVersion,
+				}
+			}
+		}
+	}
+
+	rs, err := c.recommLister.Recommendations(req.Namespace).List(labels.Everything())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	rm := map[string]*analysisv1alph1.Recommendation{}
+	for _, r := range rs {
+		k := objRefKey(r.Spec.TargetRef.Kind, r.Spec.TargetRef.APIVersion, r.Spec.TargetRef.Name)
+		rm[k] = r.DeepCopy()
+	}
+
+	var refs []corev1.ObjectReference
+
+	for k, id := range identities {
+		if r, exists := rm[k]; exists {
+			refs = append(refs, corev1.ObjectReference{
+				Kind:       rm[k].Kind,
+				Name:       rm[k].Name,
+				APIVersion: rm[k].APIVersion,
+				UID:        rm[k].UID,
+			})
+			found := false
+			for _, or := range r.OwnerReferences {
+				if or.Name == id.Name && or.Kind == id.Kind && or.APIVersion == id.APIVersion {
+					found = true
+					break
+				}
+			}
+			if !found {
+				nr := r.DeepCopy()
+				nr.OwnerReferences = append(nr.OwnerReferences,
+					*metav1.NewControllerRef(a, schema.GroupVersionKind{Version: a.APIVersion, Kind: a.Kind}))
+				if err = c.Update(ctx, nr); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		} else {
+			if err = c.createRecommendation(ctx, a, id, &refs); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+	na := a.DeepCopy()
+	na.Status.Recommendations = refs
+	t := metav1.Now()
+	na.Status.LastSuccessfulTime = &t
+	if err = c.Client.Update(ctx, na); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if a.Spec.CompletionStrategy.CompletionStrategyType == analysisv1alph1.CompletionStrategyPeriodical {
+		if a.Spec.CompletionStrategy.PeriodSeconds != nil {
+			d := time.Second * time.Duration(*a.Spec.CompletionStrategy.PeriodSeconds)
+			c.Logger.V(5).Info("Will re-sync", "after", d)
+			return ctrl.Result{
+				RequeueAfter: d,
+			}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (ac *Controller) createRecommendation(ctx context.Context, a *analysisv1alph1.Analytics,
+	id ObjectIdentity, refs *[]corev1.ObjectReference) error {
+
+	r := &analysisv1alph1.Recommendation{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-%s-%s-%s",
+				a.Name, strings.ToLower(id.Kind), strings.ReplaceAll(id.APIVersion, "/", "-"), id.Name),
```

Comment:
should we put apiversion to recommendation name?
use a.Name as generate name if good enough to me